### ISSUE TITLE
Make title block in ConTeXt template more flexibel

### ISSUE THREAD

**Issue by denismaier** [jgm/pandoc#5952]
Currently, if we go to or through ConTeXt, the way the title block gets formatted is hard coded in the template. I'm thinking about making this more flexible through variables and setups. (I have something like [this](https://tex.stackexchange.com/a/83769) in mind.) Would that be a welcome enhancement?

**Comment by jgm** [comment]
It's hard to say in the abstract.  If you make a concrete proposal I'd take a look. In general, there comes a point where saying "just use a custom template" makes more sense than making the default templates more complex.

**Comment by denismaier** [comment]
Ok. A first step would be this.

1. Assign `author`, `title` etc. to variables like so:

```
\setvariables [titleblock] [
	author={John Doe},
	title={A very interesting article},
	]
```

2. Define how this should be rendered:

```
\startsetups titleblock
  \startalignment[middle]
    {\tfd\setupinterlinespace $title$}
    $if(subtitle)$
      \smallskip
      {\tfa\setupinterlinespace $subtitle$}
    $endif$
    $if(author)$
      \smallskip
      {\tfa\setupinterlinespace $for(author)$$author$$sep$\crlf $endfor$}
    $endif$
    $if(date)$
      \smallskip
      {\tfa\setupinterlinespace $date$}
    $endif$
   \bigskip
  \stopalignment
\stopsetups
```

3. Use `\directsetup{titleblock}` to render the title block.

This would give us the same visual results than the current standard template, but users could redefine `titleblock` via `header-includes`.

**Comment by jgm** [comment]
Seems like a good idea to me.  Do any other context users have comments?

**Comment by denismaier** [comment]
Will having conditionals in code included via `header-include` work? Like, will it be possible to have 

```
$if(subtitle)$
      \smallskip
      {\tfa\setupinterlinespace $subtitle$}
```
in `header-include`?

**Comment by jgm** [comment]
Yes, that can work -- if the variable `subtitle` is empty, this section will be skipped.

**Comment by denismaier** [comment]
Hmm, I've tried it, but it looks like that everything included via `-H` comes through literally. I ended up with `$title$` instead of with the actual title.

**Comment by denismaier** [comment]
That's not a big deal, as all of these tests can also be done on the ConTeXt end. But it looks like you cannot just move the definitions from a template to a file that is included via `-H`.

**Comment by jgm** [comment]
Oh, I misunderstood what you were trying to do.  What you want is a template partial. See the documentation for instructions.  You can't do this via header-includes, but you can include a partial in the template.

**Comment by denismaier** [comment]
Yeah, a partial would be the way to go. But the idea is that users are able to override the default settings. What would be the best way to do this? 
What about a new command line option for partials? `-P`
Or perhaps each template could have a `${ user-config() }`? (What happens if a partial does not exists? Will pandoc ignore this or will it throw errors?

**Comment by jgm** [comment]
You can put variables (which the user can set) inside the partial. The partial can be set up so that it does nothing at all unless a variable is set.

**Comment by jgm** [comment]
Pandoc will throw an error if the partial doesn't exist, yes.

**Comment by denismaier** [comment]
> You can put variables (which the user can set) inside the partial.

Yes. But my goal is this: I define a setup in the template and users can override the definition in a reusable way. So they will need to have a setup that checks if certain variables are present or not. I guess the best, and currently only possible way, is to do this on the ConTeXt end.

**Comment by jgm** [comment]
Well, you can define a partial for this part, say titleblock.context.  This could go in the `data/templates` directory and the context template could refer to it (see `styles.html` for an example).  And then users can override the partial by putting a custom `templates/titleblock.context` in their user data directory.  Note that this method would change the default for all the user's documents, not just for one document, so maybe it's not what you're looking for.  I guess that's why you were wondering whether we should have a `--partial` command line option to allow specifying a particular partial on a per-document basis?

**Comment by denismaier** [comment]
You nailed it. That's exactly what I was thinking about. Basically, an option like `-H`, but it should resolve pandoc's conditionals before assembling the file to be processed.

**Comment by jgm** [comment]
It wouldn't be exactly like `-H`, because it would just say "use this partial IF the template calls for a partial with this name."  It wouldn't cause anything to be inserted unless the template requested the partial.

**Comment by tarleb** [comment]
Do I understand correctly that this boils down to a feature request for jgm/doctemplates? Should the issue be moved there?

**Comment by jgm** [comment]
The request for a change to the context template still belongs here, I think.
(As does the request for an additional command-line option.)
It may be that implementing these requests (if we decided to) would require additional changes in doctemplates.
